### ISSUE TITLE
Switching service selector causes small amount of downtime with NEGs

### ISSUE THREAD

**Issue by inversion** [kubernetes/ingress-gce#638]
I have implemented a blue/green update strategy for my deployment by updating a version label in the deployment template. I then update the service selector to use the new version label when enough pods of the new version are ready.

I have found that strategy this causes about 5-10 seconds of downtime (HTTP 502s) when using NEGs. The service events show:

```
Detach 2 network endpoint(s) (NEG "k8s1-my-service-neg" in zone "us-central1-b")
... 2 second delay...
Attach 1 network endpoint(s) (NEG "k8s1-my-service-neg" in zone "us-central1-b")
```

I'm wondering if there is a way to avoid changing service selectors causing downtime?

I suspect this issue is related to
https://cloud.google.com/kubernetes-engine/docs/how-to/container-native-load-balancing#scale-to-zero_workloads_interruption (#583)

**Comment by rramkumar1** [comment]
@freehan Any thoughts here?

**Comment by freehan** [comment]
Can you allow some period of overlap during transition? 

For instance, you have blue pods and green pods. Suppose the service points to blue pods initially.
Change the service selector to include both blue and green pods. 
Wait for some time and make sure both are taking traffic. 
Then change the service selector again to only target green pods. 

This should help. Please let us know if this still causes downtime.

**Comment by rubensayshi** [comment]
@freehan the point of a blue/green deploy is to not have any overlap at all.

I'm still facing this issue, with NEG there's a few seconds of 502s and without NEG (using a Ingress with NodePort service) it continues to send some requests to the old pods for a while :/

**Comment by rubensayshi** [comment]
@freehan @bowei I'm still facing this issue, not sure how to best revive this discussion?

the point of a blue/green deploy is to not have any overlap at all.

with NEG there's a few seconds of 502s when changing the selector.

without NEG (ofc in that case using a NodePort service) it continues to send some requests to the old pods for quite some time ...

the only thing that actually works properly is a ClusterIP service without an ingress, in which case it switches perfectly.  
but unfortunately that's not enough because we need an ingress :D

**Comment by rubensayshi** [comment]
it seems like the issue is really that when the new endpoints are added to the NEG, their health checks are UNKNOWN for a few seconds (regardless if the pods already exist / are read or not) and it will return 502s for those ~5 seconds?

**Comment by freehan** [comment]
@rubensayshi  Yes. You pointed out the key gap. The 502s are caused by the race between 2 parallel operations: 1. old endpoints getting removed 2. new endpoints getting added.  If 2 is slower than 1, then you ended up with 502s during the transitions. 

I would suggest replacing `sleep X`[here](https://github.com/kubernetes/ingress-gce/issues/638#issuecomment-646499539) with something slightly more sophisticated. For instance, validating if the `v2` app is taking traffic and performing as expected. Then the rollout system can decide whether roll back or proceed roll out. Hence the workflow looks like:

```
app: myapp; version: v1 -> app: myapp -> monitor/validation -> app: myapp; version: v2 -> monitor/validation
```

FYI, we are building out something that would allow a more seamless transition for blue/green deployments. That is going to exposed in a higher level API hence one would not need to tweak the service selector.

**Comment by rubensayshi** [comment]
hey @freehan if there's something better on the horizon then I think we'll settle for the almost-perfect solution with the `sleep`, and ye the sleep was just a rough example, I already put the version in the response header and just ping the endpoint until I get 100% of the new version :D.

is that "something better" to be expected in the near future? end of Q2 or begin of Q3? and how can I make sure I won't miss the release of that, will it be listed in the GKE update notes? https://cloud.google.com/kubernetes-engine/docs/release-notes

**Comment by haizaar** [comment]
Google acknowledge they have an issue and there is a public bug we can track: https://issuetracker.google.com/issues/180490128

**Comment by rubensayshi** [comment]
nice, thanks for sharing @haizaar , I'll subscribe to that issue.

**Comment by haizaar** [comment]
Had a chance to talk to GKE PM regarding this problem. While they won't fix the current issue (which is caused by NEG switching), they do have plans to alleviate it by supporting [K8s Gateway API](https://gateway-api.sigs.k8s.io/), which is really great since we would be able to do at least basic traffic management and routing without bringing heavy-weight likes of Istio.

If you are on GKE, reach out to your GCP account manager to join a private preview of this feature once it's available. Or use other K8s Gateway implementations that already implement it.

/CC @Keidrych

**Comment by mwarkentin** [comment]
We ran into this as well recently, wanted to provide the response I got from our GCP account team:

> This seems to be an expected behavior with the current setup, essentially a race condition between the NEG controller and the Load Balancer's health checks.
When you update the selector on the Service to point to the "Green" pods, the NEG controller effectively flushes the "Blue" IPs and populates the NEG with the "Green" IPs instantly. So, when the new endpoints get introduced to the GLB, the GLB treats them as unhealthy/unknown endpoints until they pass the GLB's specific health check and completes the hand-off. So, even if the pods are ready, it will take a while until the health check is completed.

> Essentially, there's a workaround without having to use the Gateway API for this:
This workaround functions by moving the "switch" logic from the K8S Service layer up to the LB layer. You would create a separate Service for your "Green" deployment, which generates a distinct Standalone NEG, and then attach this new NEG to your existing GLB Backend Service alongside the "Blue" NEG. Because the LB respects health checks, it will continue routing traffic exclusively to the healthy Blue NEG while the Green NEG warms up. once the Green endpoints are verified as healthy by the LB, you can safely detach the Blue NEG, ensuring there is never a gap where the LB has no healthy backends to serve traffic.

> If you want Argo Rollouts to drive the switch from within the cluster and maintain that Kubernetes-native feel, then Gateway API is the required solution.
Argo Rollouts and Gateway API Integration: [Argo Rollouts Gateway API plugin - Documentation](https://rollouts-plugin-trafficrouter-gatewayapi.readthedocs.io/en/latest/)
